### ISSUE TITLE
Uniform RV Outside of Support (pymc 2.3.3)

### ISSUE THREAD

**Issue by kyleabeauchamp** [pymc-devs/pymc3#612]
I have a simple model with some uniform random variables, a deterministic, and an observation.  I find that occasionally this leads to proposals of my uniform RVs outside of their support.  Is this the expected behavior?  Or maybe I'm doing something silly.  Below is my driver script, which occasionally dies with a ValueError due to the incorrect sign...

```
import pymc

x = pymc.Uniform("x", lower=0.0, upper=1.0)
y = pymc.Uniform("y", lower=0.0, upper=1.0)

def calc(x, y):
    if x < 0:
        print("x=%f" % x)
        raise(ValueError("Wrong sign!"))    
    return 1.0

estimator = pymc.Deterministic(eval=lambda x, y: calc(x, y), doc="Calculate", name="estimator", parents=dict(x=x, y=y))
measurement = pymc.Normal("observed", mu=estimator, tau=1.0, value=1.0, observed=True)

variables = [x, y, estimator, measurement]

mcmc = pymc.MCMC(variables)
mcmc.sample(10000)
```

Here is output:

```
    raise(ValueError("Wrong sign!"))    
ValueError: Wrong sign!
```

**Comment by fonnesbeck** [comment]
Kyle,

I'm not able to replicate this behavior, even after running it for 100K iterations. What is your setup?

BTW, `Model` is not meant to be instantiated directly. Easiest just to:

```
mcmc = pymc.MCMC(variables)
```

**Comment by kyleabeauchamp** [comment]
OK, I have updated my code to avoid the Model building step.  I still see the issue.

So I _believe_ I see the issue primarily early in the MCMC chain--running longer chains do not seem to increase the likelihood of the ValueError.  In fact, I previously used a bash script like the following to help reproduce things:

```
python test_pymc_bug.py
python test_pymc_bug.py
python test_pymc_bug.py
python test_pymc_bug.py
python test_pymc_bug.py
python test_pymc_bug.py
python test_pymc_bug.py
python test_pymc_bug.py
```

**Comment by kyleabeauchamp** [comment]
Actually, I'm not sure how long the chains need to be.

**Comment by kyleabeauchamp** [comment]
But I will say that this issue has been _very_ dependent on the "state".  For example, adding extraneous lines of code to the script seemed to change the outcome.  Perhaps this is the RNG seed, or perhaps something more sinister like memory state.

The probabily of ValueError also seems to depend on whether I run the script as a standalone python script or in IPython.

**Comment by kyleabeauchamp** [comment]
System Details:

Ubuntu 14.04 LTS 64 bit.  Intel 3770K, 16GB RAM.  Python 2.7 (Anaconda), with PyMC 2.3.3 installed via conda.

**Comment by kyleabeauchamp** [comment]
I've also replicated this on a Ubuntu 12.04 64 bit machine (again with Anaconda Python 2.7).

**Comment by fonnesbeck** [comment]
If you could isolate it with a seed, that would be great.

Can you try replicating it with a build from the current 2.3 branch? There have been a few bugs that mysteriously disappear in 2.3.4.

**Comment by kyleabeauchamp** [comment]
OK, so the latest 2.3 branch + miniconda (commands below) seem to fix the issue.  Thus, the problem is either:
1.  Fixed by a recent change
2.  Caused by the conda build.  

```
wget http://repo.continuum.io/miniconda/Miniconda-3.7.0-Linux-x86_64.sh
~/miniconda/bin/conda install numpy scipy matplotlib nose ipython
~/miniconda/bin/python setup.py install
```

It's hard for me to isolate (1) or (2), as I can't use my full Anaconda install to build the from source, due to missing Lapack libraries (e.g. full Anaconda numpy / scipy is linked against MKL, which then causes difficulties for pymc's setup.py).

**Comment by kyleabeauchamp** [comment]
On my full Anaconda build, I tried the following:

```
conda remove accelerate
conda install pymc
[...]
python test_pymc_bug.py
(Still see issue)
```

```
conda remove accelerate
python setup.py install
[...]
python test_pymc_bug.py
(No issue)
```

This leads to suspect something fishy in the conda build.  Again, I can't guarantee it because the issue is pretty hard to reproduce, so I could be reporting noise here...

**Comment by fonnesbeck** [comment]
Yeah, its high time we released 2.3.4. (#594)

**Comment by kyleabeauchamp** [comment]
Yes.  I did another test using the 2.3.3 source tarball.  The bug has been fixed somewhere between 2.3.3 and master.

To clarify, the bug _was_ present in my 2.3.3. source build, but not present in the master.  Thus, Conda is not at fault.

**Comment by kyleabeauchamp** [comment]
AFAIK this is fixed in the 2.3.4 release, including via my conda package.
